pkg/controller/cloudfront/distribution/setup.go

```diff
@@ -0,0 +1,182 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package distribution
+
+import (
+	"context"
+	"encoding/json"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/cloudfront"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/cloudfront/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// TODO: isn't this defined as an API constant somewhere in aws-sdk-go? Generated zz_enums.go seems not to contain it either
+const stateDeployed = "Deployed"
+
+// SetupDistribution adds a controller that reconciles Distribution.
+func SetupDistribution(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(svcapitypes.DistributionGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&svcapitypes.Distribution{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.DistributionGroupVersionKind),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithExternalConnecter(&connector{
+				kube: mgr.GetClient(),
+				opts: []option{
+					func(e *external) {
+						e.preCreate = preCreate
+						e.postCreate = postCreate
+						e.lateInitialize = lateInitialize
+						e.preObserve = preObserve
+						e.postObserve = postObserve
+						e.isUpToDate = isUpToDate
+						e.preUpdate = preUpdate
+						e.preDelete = preDelete
+					},
+				},
+			}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.Distribution, cdi *svcsdk.CreateDistributionInput) error {
+	cdi.DistributionConfig.CallerReference = awsclients.String(string(cr.UID))
+	cdi.DistributionConfig.Origins.Quantity =
+		awsclients.PInt64Value(len(cr.Spec.ForProvider.DistributionConfig.Origins.Items))
```

Comment:
Hi @muvaf,
Thanks for the review. 
I have removed the `aws.PInt64Value` function in favor of `aws.Int64` as suggested. 

Regarding the nil check, do you mean `cr.DistributionConfig.Origins` instead of `cdi`? The SDK type for the fields `DistributionConfig` and `Origins` both have the yaml tags `required`. Does that guarantee them to be non-nil?

Reply:
I meant `cdi.DistributionConfig.Origins` which is defined in AWS SDK and a pointer, i.e. the code in our repo sets its value, hence can be `nil`. But I also checked `cr.Spec.ForProvider.DistributionConfig.Origins` and it seems that's [not marked as `required`](https://github.com/crossplane/provider-aws/pull/654/files#diff-92efa1c153bb84bc723bdcd87cc53bb9b584ab93e4a2bf9fe6b5dd244356ea95R378) in CRD YAML that's been generated (probably because we yet to produce `required` markers for types in `types.go` file)

One way to be sure could be counting `cdi.DistributionConfig.Origins.Quantity.Items` if `cdi.DistributionConfig.Origins` is not `nil` because `Quantity` field is bound to `Items` on that struct so you'd need to have two `if`s.

Reply:
Good observation @muvaf. The assumption I made was  `cr.Spec.ForProvider.DistributionConfig.Origins` and in turn `cdi.DistributionConfig.Origins` can never be `nil` because it's marked as required in AWS API. Why don't we also mark it as required in generated code? We now check if the CR has non-nil `Origins`.